### PR TITLE
fix(transport): don't pace below timer granularity

### PR DESCRIPTION
**Pull request by mxinden**
Neqo assumes a timer granularity of 1ms:

https://github.com/mozilla/neqo/blob/0eb9174d7a67b250607f0c3ea056fe056bcf91f5/neqo-transport/src/rtt.rs#L25-L27

but `neqo_transport::Pacer::next()` might return values `< GRANULARITY`.

https://github.com/mozilla/neqo/blob/0eb9174d7a67b250607f0c3ea056fe056bcf91f5/neqo-transport/src/pace.rs#L71-L90

Under the assumption that a timer implementation rounds small values up to its granularity (e.g. 1ms), packets can be delayed significantly more than intended by `Pacer`.

With this commit `Pacer` does not delay packets that would previously be delayed by less than `GRANULARITY`. The downside is loss in pacing granularity.

See also:

- google/quiche
  - https://github.com/google/quiche/blob/60aec87316d24392b2ea37c391ecf406ef183074/quiche/quic/core/congestion_control/pacing_sender.cc#L167
  - https://github.com/google/quiche/blob/60aec87316d24392b2ea37c391ecf406ef183074/quiche/quic/core/quic_constants.h#L304
- quic-go
  - https://github.com/quic-go/quic-go/blob/d1f9af4cc6b13c96dc302ac9ec5f061ed294d36b/internal/protocol/params.go#L137
- `kGranularity` in RFC 9002 https://datatracker.ietf.org/doc/html/rfc9002#name-constants-of-interest

---

Initial idea from @larseggert.

Would fix performance regression in https://github.com/mozilla/neqo/pull/2008:

```
➜  neqo git:(pace-granularity) ✗ critcmp main main-no-pacing 32k 32k-no-pacing 32k-min-1ms-pacing -f "Download" --list
1-conn/1-100mb-resp (aka. Download)/client
------------------------------------------
32k-no-pacing          1.00    114.8±49.78ms   871.3 MB/sec
32k-min-1ms-pacing     1.04    119.9±63.28ms   834.2 MB/sec
main-no-pacing         1.05    121.0±83.97ms   826.4 MB/sec
main                   1.45    165.9±56.69ms   602.8 MB/sec
32k                   11.39  1307.8±1038.97ms   76.5 MB/sec
```

### CODE REVIEW

File: neqo-transport/src/pace.rs [hunk 1]
```diff
@@ -74,19 +76,30 @@ impl Pacer {
...
+        }
+
+        // This is the inverse of the function in `spend`:
+        // self.t + rtt * (self.p - self.c) / (PACER_SPEEDUP * cwnd)
+        let r = rtt.as_nanos();
+        let d = r.saturating_mul(u128::try_from(self.p - self.c).unwrap());
+        let add = d / u128::try_from(cwnd * PACER_SPEEDUP).unwrap();
+        let w = u64::try_from(add).map(Duration::from_nanos).unwrap_or(rtt);
+
+        // If next is below timer granularity, send immediately.
```
Comment (martinthomson) [inline neqo-transport/src/pace.rs:90]:
```suggestion
        // If the increment is below the timer granularity, send immediately.
```

File: neqo-transport/src/pace.rs [hunk 2]
```diff
@@ -74,19 +76,30 @@ impl Pacer {
...
+        if w < GRANULARITY {
+            qtrace!(
+                [self],
+                "next {cwnd}/{rtt:?} below granularity ({w:?} < {GRANULARITY:?}) no wait = {:?}",
+                self.t
```
Comment (martinthomson) [inline neqo-transport/src/pace.rs:95]:
```suggestion
                "next {cwnd}/{rtt:?} wait ({w:?} below granularity",
```

You don't need to log `self.t` here.  It's always in the past.

**Review by martinthomson — COMMENTED** [review]
Would you mind writing a test for this?

I also see a lot of CI failures here.  I don't immediately see why this is the case, likely the test was dependent on some amount of pacing happening (with a wait that might be suppressed by this change).

File: neqo-transport/src/connection/tests/migration.rs [hunk 3]
```diff
@@ -160,7 +152,6 @@ fn path_forwarding_attack() {
     assert_v6_path(&client_data2, false);
 
     // The server keeps sending on the new path.
-    now = skip_pacing(&mut server, now);
```
Comment (mxinden) [inline neqo-transport/src/connection/tests/migration.rs:163]: Pacing on new path is now below granularity and thus packet on new path is send
immediately. Calling `skip_pacing` will instead fast forward to the PTO of the
old path to expire, thus leading to an unexpected probe packet on the old path.

```
thread 'connection::tests::migration::path_forwarding_attack' panicked at test-fixture/src/assertions.rs:153:5:
assertion `left == right` failed
  left: [fe80::1]:443
 right: 192.0.2.1:443
```

3c65482953cf2b3d151176bd96eedd80e586aa83 simply removes the no longer needed `skip_pacing` step, thus
reverting to the previous behavior.

**Comment by mxinden** [comment]
> Would you mind writing a test for this?

Of course. Thanks for continuously pushing for more tests @martinthomson. Sorry for not providing one early on.

https://github.com/mozilla/neqo/pull/2035/commits/f183df88e8ff4a475c16deb4225c37777e126736 adds a basic test. Let me know if you would like me to cover more scenarios.

Also, this might be a good opportunity for property based testing e.g. via [`quickcheck`](https://docs.rs/quickcheck/latest/quickcheck/). Let me know if that sounds worth the additional dependency. Happy to add it, preferably in a separate pull request.

**Comment by mxinden** [comment]
> ## [Codecov](https://app.codecov.io/gh/mozilla/neqo/pull/2035?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mozilla) Report
> 
> Attention: Patch coverage is `96.55172%` with `1 line` in your changes missing coverage. Please review.
> 
> > Project coverage is 95.05%. Comparing base [(`0eb9174`)](https://app.codecov.io/gh/mozilla/neqo/commit/0eb9174d7a67b250607f0c3ea056fe056bcf91f5?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mozilla) to head [(`c66154b`)](https://app.codecov.io/gh/mozilla/neqo/commit/c66154b1b28f9e3044e49c20ca002d0db1a14799?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mozilla).
> > Report is 2 commits behind head on main.
> 
> [Files](https://app.codecov.io/gh/mozilla/neqo/pull/2035?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mozilla) 	Patch % 	Lines
> [neqo-transport/src/pace.rs](https://app.codecov.io/gh/mozilla/neqo/pull/2035?src=pr&el=tree&filepath=neqo-transport%2Fsrc%2Fpace.rs&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mozilla#diff-bmVxby10cmFuc3BvcnQvc3JjL3BhY2UucnM=) 	96.55% 	[1 Missing ⚠️ ](https://app.codecov.io/gh/mozilla/neqo/pull/2035?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mozilla)
> Additional details and impacted files
> 
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/mozilla/neqo/pull/2035?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mozilla). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mozilla).

I am surprised that Codecov reports a line in a unit test as missing test coverage:

image

**Review by larseggert — APPROVED** [review]

**Pull request merged**